File: jscode/src/modules/websocket.ts
import WebSocket from 'ws';
import fs from 'fs';
import yaml from 'js-yaml';

/**
 * Class representing a WebSocket connection.
 */
class cbws {
    websocket: WebSocket;

    /**
     * Constructs a new cbws instance and initializes the WebSocket connection.
     */
    constructor() {
        const uniqueConnectionId = this.getUniqueConnectionId();
        const initialMessage = this.getInitialMessage();
   
        const agentIdParam = process.env.agentId ? `&agentId=${process.env.agentId}` : '';
        const parentIdParam = process.env.parentId ? `&parentId=${process.env.parentId}` : '';
        this.websocket = new WebSocket(`ws://localhost:${process.env.SOCKET_PORT}/codebolt?id=${uniqueConnectionId}${agentIdParam}${parentIdParam}${process.env.Is_Dev ? '&dev=true' : ''}`);
        this.initializeWebSocket(initialMessage).catch(error => {
            console.error("WebSocket connection failed:", error);
        });
    }
    private getUniqueConnectionId(): string {
        try {
            let fileContents = fs.readFileSync('./codeboltagent.yaml', 'utf8');
            let data: any = yaml.load(fileContents);
            return data.unique_connectionid;
        } catch (e) {
            console.error('Unable to locate codeboltagent.yaml file.');
            return '';
        }
    }

    private getInitialMessage(): string {
        try {
            let fileContents = fs.readFileSync('./codeboltagent.yaml', 'utf8');
            let data: any = yaml.load(fileContents);
            return data.initial_message;
        } catch (e) {
            console.error('Unable to locate codeboltagent.yaml file.');
            return '';
        }
    }

    /**
     * Initializes the WebSocket by setting up event listeners and returning a promise that resolves
     * when the WebSocket connection is successfully opened.
     * @returns {Promise<WebSocket>} A promise that resolves with the WebSocket instance.
     */
    private async initializeWebSocket(initialMessage: string): Promise<WebSocket> {
        return new Promise((resolve, reject) => {
            this.websocket.on('error', (error: Error) => {
                console.log('WebSocket error:', error);
                reject(error);
            });

            this.websocket.on('open', () => {
                console.log('WebSocket connected');
                // if (this.websocket) {
                //     this.websocket.send(JSON.stringify({
                //         "type": "sendMessage",
                //         "message": initialMessage
                //     }));
                //     resolve(this.websocket);
                // }
            });

            this.websocket.on('message', (data: WebSocket.Data) => {
                // Handle incoming WebSocket messages here.
                // console.log('WebSocket message received:', data);
            });
        });
    }

    /**
     * Getter for the WebSocket instance. Throws an error if the WebSocket is not open.
     * @returns {WebSocket} The WebSocket instance.
     * @throws {Error} If the WebSocket is not open.
     */
    get getWebsocket(): WebSocket {
        if (!this.websocket.OPEN) {
            throw new Error('WebSocket is not open');
        } else {
            return this.websocket;
        }
    }
}

export default new cbws();

File: modules/websocket.py
import asyncio
import websockets
import yaml
import os

class CBWS:
    def __init__(self):
        self.websocket = None
        self.unique_connection_id = self.get_unique_connection_id()
        self.initial_message = self.get_initial_message()
        self.loop = asyncio.get_event_loop()
        self.loop.run_until_complete(self.initialize_websocket())

    def get_unique_connection_id(self):
        try:
            with open('./codeboltagent.yaml', 'r') as file:
                data = yaml.safe_load(file)
                return data.get('unique_connectionid', '')
        except FileNotFoundError:
            print('Unable to locate codeboltagent.yaml file.')
            return ''

    def get_initial_message(self):
        try:
            with open('./codeboltagent.yaml', 'r') as file:
                data = yaml.safe_load(file)
                return data.get('initial_message', '')
        except FileNotFoundError:
            print('Unable to locate codeboltagent.yaml file.')
            return ''

    async def initialize_websocket(self):
        agent_id_param = f"&agentId={os.getenv('agentId')}" if os.getenv('agentId') else ''
        parent_id_param = f"&parentId={os.getenv('parentId')}" if os.getenv('parentId') else ''
        dev_param = '&dev=true' if os.getenv('Is_Dev') else ''
        uri = f"ws://localhost:{os.getenv('SOCKET_PORT')}/codebolt?id={self.unique_connection_id}{agent_id_param}{parent_id_param}{dev_param}"

        try:
            async with websockets.connect(uri) as websocket:
                self.websocket = websocket
                print('WebSocket connected')
                # Uncomment the following lines to send the initial message
                # await websocket.send(json.dumps({
                #     "type": "sendMessage",
                #     "message": self.initial_message
                # }))
                # Handle incoming messages
                async for message in websocket:
                    # Handle incoming WebSocket messages here.
                    # print('WebSocket message received:', message)
                    pass
        except Exception as e:
            print("WebSocket connection failed:", e)

    def get_websocket(self):
        if not self.websocket or self.websocket.closed:
            raise Exception('WebSocket is not open')
        return self.websocket

# Create an instance of the CBWS class
cbws_instance = CBWS()
